Kore Alpha Bootstrap (Binaries)

gambol99 opened the pull request:
Given we are installing applications now and require the ability to run workloads we can no longer rely on docker-compose for the local dev or demo. Part of this also feeds into the requirement of bootstrapping a production kore installed as well. I've added the alpha bootstrap command. This version pulls the helm chart down from the release, though we could bindata the directory and compile it inside instead.

I hope we can extend via the UI and service to address the long running story https://github.com/appvia/kore/issues/340 and use the UI as a wizard and a service to install a production grade kore installation

### Testing

Bring up a local installation
```shell
$ kore alpha bootstrap up 
✅ Passed preflight checks for deployment
✅ Passed preflight checks for local cluster provider
✅ Attempting to build the local kubernetes cluster
   ◉ Checking if kind cluster: "kore" already exists
   ◉ Kind cluster: "kore" already exists, skipping creation
   ◉ Ensuring the kind cluster: "kore" is running
   ◉ Exporting kubeconfig from kind cluster: "kore"
✅ Exported the kubeconfig from provisioned cluster
✅ Provisioned a local kubernetes cluster
✅ Configured the kubectl context: kind-kore
✅ Attempting to deploy the Kore release v0.2.0-rc2
✅ Deployed the Kore release into the cluster
✅ Waiting for deployment to rollout successfully (5m0s timeout)
✅ Successfully deployed the kore release to cluster

You can access the Kore portal via http://localhost:3000
Configure your CLI via $ kore login -a http://localhost:10080 local
```
You can destroy via `kore alpha bootstrap destroy` ... shutdown or pause via `kore alpha bootstrap stop` and bring back up (state included) via `kore alpha bootstrap up` again

You can override the release chart by using `--release` i.e. `kore alpha bootstrap up --release ./charts/kore`

Assuming none of the binaries exists
```shell
$ kore alpha bootstrap up 
✅ Passed preflight checks for deployment
   ◉ Kind binary not found in $PATH
   ◉ Download: https://github.com/kubernetes-sigs/kind/releases/download/v0.8.1/kind-linux-amd64 (/home/jest/bin/kind) (y/N)? y
   ◉ Attempting to download the kind binary
✅ Passed preflight checks for local cluster provider
✅ Attempting to build the local kubernetes cluster
   ◉ Checking if kind cluster: "kore" already exists
   ◉ Provisioning a kind cluster: "kore" (usually takes 3-5mins)
   ◉ Exporting kubeconfig from kind cluster: "kore"
✅ Exported the kubeconfig from provisioned cluster
✅ Provisioned a local kubernetes cluster
✅ Configured the kubectl context: kind-kore
   ◉ Helm binary not found in $PATH
   ◉ Download https://get.helm.sh/helm-v3.2.1-linux-amd64.tar.gz (/home/jest/bin/helm) y/N? y
✅ Attempting to download helm release from github
✅ Downloaded the helm release (/home/jest/bin/helm)
✅ Attempting to deploy the Kore release v0.2.0-rc2
✅ Deployed the Kore release into the cluster
✅ Waiting for deployment to rollout successfully (5m0s timeout)
✅ Successfully deployed the kore release to cluster
```

Replaces #857

mrsheepuk commented:
One question on the 'download binaries' approach (which seems like a great idea), if we download the kind and helm binaries, should we put them in the `.kore` directory so they're specific to our application, instead of the user's bin directory?

gambol99 commented:
> One question on the 'download binaries' approach (which seems like a great idea), if we download the kind and helm binaries, should we put them in the `.kore` directory so they're specific to our application, instead of the user's bin directory?

```shell
✅ Passed preflight checks for deployment
   ◉ Kind binary not found in $PATH
   ◉ Download: https://github.com/kubernetes-sigs/kind/releases/download/v0.8.1/kind-linux-amd64 (/home/jest/.kore/build/kind) (y/N)? y
   ◉ Attempting to download the kind binary
✅ Passed preflight checks for local cluster provider
✅ Attempting to build the local kubernetes cluster
   ◉ Checking if kind cluster: "kore" already exists
   ◉ Kind cluster: "kore" already exists, skipping creation
   ◉ Ensuring the kind cluster: "kore" is running
   ◉ Exporting kubeconfig from kind cluster: "kore"
✅ Exported the kubeconfig from provisioned cluster
✅ Provisioned a local kubernetes cluster
✅ Configured the kubectl context: kind-kore
   ◉ Helm binary not found in $PATH
   ◉ Download https://get.helm.sh/helm-v3.2.1-linux-amd64.tar.gz (/home/jest/.kore/build/helm) y/N? y
✅ Attempting to download helm release from github
✅ Downloaded the helm release (/home/jest/.kore/build/helm)
✅ Attempting to deploy the Kore release v0.2.0-rc2
✅ Deployed the Kore release into the cluster
✅ Waiting for deployment to rollout successfully (5m0s timeout)
```

Move the download directory into the KORE_CONFIG path

bandesz commented:
There is a conflict which needs to be resolved.

Also I think the command name "bootstrap" is not right, as e.g. `kore alpha bootstrap destroy` doesn't make much sense. I think we should use local:
```
$ kore alpha local up
$ kore alpha local stop
$ kore alpha local destroy
```

bandesz commented on deploy/kube/portal.yml:
```diff
@@ -15,12 +15,15 @@ spec:
     name: portal
 ---
-apiVersion: extensions/v1beta1
+apiVersion: apps/v1
```
I'm not sure these are still up-to-date. Are we still using these or should we delete them?

gambol99 replied:
Indeed, removed https://github.com/appvia/kore/pull/871/commits/f98387b86a3d998679612609a6249dbbb39409a1

bandesz commented on pkg/cmd/kore/bootstrap/bootstrap.go:
```diff
@@ -0,0 +1,116 @@
+var (
+	usage = `
+Bootstrap provides an experimental means of bootstrapping a local Kore installation. At
+present the local installation use "kind" https://github.com/kubernetes-sigs/kind.
```
use -> uses

bandesz commented on pkg/cmd/kore/bootstrap/bootstrap.go:
```diff
@@ -0,0 +1,116 @@
+
+Unless specified otherwise it will deploy an official tagged release from Github, though
+this can be overriden using the --release flag. Note the installation is performed via
+helm with a local values.yaml is generated in the directory; so if you wish to change
```
`is generated` -> `generated`

bandesz commented on pkg/cmd/kore/bootstrap/bootstrap.go:
```diff
@@ -0,0 +1,116 @@
+Unless specified otherwise it will deploy an official tagged release from Github, though
+this can be overriden using the --release flag. Note the installation is performed via
+helm with a local values.yaml is generated in the directory; so if you wish to change
+any of the values post a installation can change this file and run the 'up' command.
```
`post a installation can change this file and run the 'up' command.` -> `post installation, you can change this file and run the 'up' command.`

bandesz commented on pkg/cmd/kore/bootstrap/bootstrap.go:
```diff
@@ -0,0 +1,116 @@
+helm with a local values.yaml is generated in the directory; so if you wish to change
+any of the values post a installation can change this file and run the 'up' command.
+
+Note, the local installtions data persistency is tied to the provider. For kind as long
```
`installtions` -> `installation`

bandesz commented on pkg/cmd/kore/bootstrap/bootstrap.go:
```diff
@@ -0,0 +1,116 @@
+any of the values post a installation can change this file and run the 'up' command.
+
+Note, the local installtions data persistency is tied to the provider. For kind as long
+as the container is not delete the data is kept.
```
`delete` -> `deleted`

bandesz commented on pkg/cmd/kore/bootstrap/bootstrap.go:
```diff
@@ -0,0 +1,116 @@
+$ kore alpha bootstrap up --release ./charts
+$ kore alpha bootstrap up --release https://URL
+
+# Destroy the local installtion
```
`installtion` -> `installation`

bandesz commented on pkg/cmd/kore/bootstrap/bootstrap.go:
```diff
@@ -0,0 +1,116 @@
+# Destroy the local installtion
+$ kore alpha bootstrap destroy
+
+# To stop the local installed without deleting the data
```
`installed` -> `installation`

bandesz commented on pkg/cmd/kore/bootstrap/bootstrap.go:
```diff
@@ -0,0 +1,116 @@
+# To stop the local installed without deleting the data
+$ kore alpha bootstrap stop
+
+The application should be available on http://127.0.0.1:3000. You can provision the
```
Provisioning the CLI sounds strange. Maybe `you can set up your CLI profile`

bandesz commented on pkg/cmd/kore/bootstrap/helm.go:
```diff
@@ -0,0 +1,96 @@
+	"gopkg.in/yaml.v2"
+)
+
+// GetHelmValues returns returns or prompts for the values
```
`returns returns`

bandesz commented on pkg/cmd/kore/bootstrap/providers/kind/kind.go:
```diff
@@ -0,0 +1,253 @@
+			return err
+		}
+
+		if err := os.Chmod(path, os.FileMode(0555)); err != nil {
```
Why 0555? The user should be able to write the file. Set it to 0755.

bandesz replied:
This wasn't addressed

bandesz commented on pkg/utils/confirm.go:
```diff
@@ -0,0 +1,44 @@
+
+// AskForConfirmation asks a question of the user
+func AskForConfirmation(stdin io.Reader) bool {
+	reader := bufio.NewReader(stdin)
```
bufio doesn't handle backspaces, use the library we already use: cmd/utils/prompts.go

gambol99 replied:
Yep, this was the first thing i tried to do .. the problem was I wasn't able to keep the same formatting and so the output got all screwy as `github.com/manifoldco/promptui` fiddles with the output even when I overload the formatter and template.

bandesz commented on pkg/utils/json.go:
```diff
@@ -21,9 +21,30 @@ import (
+		return nil, err
+	}
+
+	patch, err := jsonpatch.CreateMergePatch(o, s)
```
I think this is the wrong order, as CreateMergePatch expects `original, modified`, or maybe it's confusing what src and dst are

gambol99 replied:
Deleted as its no longer required

bandesz commented on pkg/utils/yaml.go:
```diff
@@ -0,0 +1,35 @@
+)
+
+// ToYAML marshalls the struct to yaml
+func ToYAML(in interface{}) ([]byte, error) {
```
`yaml.Marshal` does exactly this, if I understand correctly

gambol99 replied:
This is largely to ensure I use the same lib every time ..

bandesz commented on .gitignore:
```diff
@@ -2,6 +2,7 @@
 .idea
 1
 /bin/
+build/
```
is this still needed?

bandesz commented on go.mod:
```diff
@@ -14,6 +14,7 @@ require (
 	github.com/blevesearch/go-porterstemmer v1.0.2 // indirect
 	github.com/blevesearch/segment v0.0.0-20160915185041-762005e7a34f // indirect
 	github.com/client9/misspell v0.3.4
+	github.com/containerd/continuity v0.0.0-20200107194136-26c1120b8d41 // indirect
```
it seems to me we don't need any of the vendor changes, do we?

gambol99 commented:
@bandesz ... fixed the various issues in https://github.com/appvia/kore/pull/871/commits/f258caaf3a55fb795e77819c17d1fe4ac6c9aef3

- fixed up the various spelling mistakes
- removed the MergeJSON method as it no longer used
- removed the github.com/containerd/continuity package from imports
- removed the build/ from .gitignore file
- moved the values.yaml to /home/jest/.kore/values.yaml
- checking if the helm binary is version v3

bandesz commented:
I ran `kore alpha local stop` then `kore alpha local up`. I've got this error:

```
❌ Deployed the Kore release into the cluster
Error: Post https://127.0.0.1:53756/api/v1/namespaces: EOF
``` 

When running `kore alpha local up` the second time, it worked without errors.

gambol99 commented:
> I ran `kore alpha local stop` then `kore alpha local up`. I've got this error:
> 
> ```
> ❌ Deployed the Kore release into the cluster
> Error: Post https://127.0.0.1:53756/api/v1/namespaces: EOF
> ```
> 
> When running `kore alpha local up` the second time, it worked without errors.

Yeah .. i think it's a timing issue, needs to wait for the kubernetes api to come up fully before helming .. https://github.com/appvia/kore/pull/871/commits/20d062fdde6c3ef85bc1c2b1c9d562461591216d
